Add static as keyword to find_library

NickeZ opened the pull request:
This patch allows the build file author to enforce static or dynamic linkage. I've found this to be necessary when the two versions are not equivalent.

AFAIK there are not really any naming standards when it comes to libraries on Windows. But from my experience it is more common to prefix the static version with 'lib' and not have any prefixes for the dynamic library. This patch therefore changes the order of the prefixes so that dynamic libraries are found first which seem to be the case already on other platforms.

tp-m commented:
Would it make sense to have different keywords like `prefer_static` and `require_static` instead of just `static`?

NickeZ commented:
Yeah I guess there are four combinations

* require static (fail if static not found)
* require shared (fail if shared not found)
* prefer static (shared is fallback) <- This is currently default on windows.
* prefer shared (static is fallback) <- This is currently default on non-windows

I don't think "require anything" can be implemented on windows because there are no ways to determine if the found library is static or shared.

An additional thing. I noted that meson itself produces lib\<name\>.a for static libraries on windows and \<name\>.dll for shared libraries. Which is an additional reason to have the right priority on the prefixes.

ignatenkobrain commented on mesonbuild/compilers/c.py:
```diff
@@ -839,6 +842,23 @@ def get_library_naming(self, env, libtype, strict=False):
+    def verify_libtype(self, filename, libtype):
+        # FIXME: shared-static and static-shared return the same for now
+        if libtype in ('default', 'shared-static', 'static-shared'):
+            return True;
```
__[Flake8]__

[E703] statement ends with a semicolon


<!-- Comment by [Sider](https://sider.review) -->

jon-turney commented on mesonbuild/compilers/c.py:
```diff
@@ -839,11 +843,29 @@ def get_library_naming(self, env, libtype, strict=False):
+            if shlibname in stdout:
+                return True
+            return False
+        raise AssertionError('BUG: unkonwn libtype: {!r}'.format(libtype))
```
unkonwn -> unknown

jon-turney commented on mesonbuild/compilers/c.py:
```diff
@@ -839,11 +843,29 @@ def get_library_naming(self, env, libtype, strict=False):
+        if libtype == 'static':
+            if shlibname in stdout:
+                return False
+            return True
```
I think you could just write `return (shlibname not in stdout)` here, and likewise the negation of that below...

jon-turney commented on mesonbuild/compilers/c.py:
```diff
@@ -839,11 +843,29 @@ def get_library_naming(self, env, libtype, strict=False):
+        # FIXME: shared-static and static-shared return the same for now
+        if libtype in ('shared-static', 'static-shared'):
+            return True
+        proc = subprocess.Popen(['lib', '/list', filename], stdout=subprocess.PIPE)
```
I'm not sure how safe the assumption that `lib` is in the path is.  Perhaps the potential exception raised if it can't be found should be handled.

I guess there is no possibility of a cross-tool being required.

nirbheek commented:
This conflicted directly with https://github.com/mesonbuild/meson/pull/3691, so it needs a rebase. I will review it after that.

ignatenkobrain commented on mesonbuild/compilers/c.py:
```diff
@@ -839,6 +843,20 @@ def get_library_naming(self, env, libtype, strict=False):
+        stdout = stdout.decode('utf-8')
+        shlibname = filename.with_suffix('.dll').name
+        if libtype == 'static':
+            return not shlibname in stdout
```
__[Flake8]__

[E713] test for membership should be 'not in'


<!-- Comment by [Sider](https://sider.review) -->

NickeZ commented:
Thanks for the comments. I've updated and rebased. Ready for review @nirbheek

NickeZ commented:
@nirbheek Fixed test failure

NickeZ commented:
ping!

NickeZ commented:
Before rebasing I would like to know if the PR is desired or not.

NickeZ commented:
rebased

xclaesse commented:
Sorry to be late on this conversation, but I personally don't like that "static-shared" stuff. I think we should only have boolean `static` keyword argument, and then we should have fallbacks:

```meson
lib = find_library('foo', static : true, required : false)
if not lib.found()
  lib = find_library('foo')
endif
```

Of course having to code that if-block is a bit ugly, that that can be part of my `dependency()` fallbacks proposal: https://github.com/mesonbuild/meson/pull/4595.

jpakkane commented:
As a feature we want this. Exactly what UI it should have is not fixed yet.

tp-m commented:
> Would it make sense to have different keywords like `prefer_static` and `require_static` instead of just `static`?

Or maybe `static: true|false` to just check for one variant and `prefer_static: true|false` to check for both but prefer static?

To me that's clearer than the string-based approach, but if we ever need to support more variants it'd be messy :)

NickeZ commented:
Actually I don't really see the reason for "prefer static/shared". IMO, since this is an option that will be written into the build file, It should either let the end-user decide (by setting default-library) or enforce static or dynamic.

What would be the use case for *preferring* static or dynamic?

I still think it is reasonable for the end-user to be able to say "prefer X but use Y if needed" for the whole build. So that stuff actually builds if you have static or dynamic libs installed.

TheQwertiest commented:
@NickeZ any plans to follow through on this PR? I can take over if you'd like.

NickeZ commented:
> As a feature we want this. Exactly what UI it should have is not fixed yet.

This PR needs feedback from the maintainers before any progress can be made. It is a waste of time to implement alternative UIs before we have some sort of consensus of what a good UI is.

nirbheek commented:
I think we should implement `static: true|false` for now and we can extend it to more things later as needed.

NickeZ commented:
> I think we should implement `static: true|false` for now and we can extend it to more things later as needed.

With what behavior? Forcing static or preferring static?

nirbheek commented:
> With what behavior? Forcing static or preferring static?

Force static.

NickeZ commented:
> Force static.

And false forces dynamic?

nirbheek commented:
> And false forces dynamic?

Yes, omitting the kwarg will keep the current behaviour.

NickeZ commented:
OK, i've rebased and changed the kwarg to static. There are changes in this PR that are not needed to actually merge the static-kwarg feature. Do you want me to move any such changes to another PR?

nirbheek commented:
That's always a good idea for minimizing delays in getting a PR merged. Small PRs are easy to review and merge :)

jpakkane commented:
Have not looked at the implementation but featurewise this is ok.

NickeZ commented:
Tests passes on my machine now. Waiting for review.

NickeZ commented:
One failure seems related. Is it on purpose that it links statically to `libm`? That seems wrong to me. I'm not sure what goes wrong here actually..

```

======================================================================
ERROR: test_pkgconfig_static (__main__.AllPlatformTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "run_unittests.py", line 144, in wrapped
    return f(*args, **kwargs)
  File "run_unittests.py", line 2401, in test_pkgconfig_static
    self.build()
  File "run_unittests.py", line 1208, in build
    return self._run(self.build_command + args + extra_args, workdir=self.builddir)
  File "run_unittests.py", line 1156, in _run
    raise subprocess.CalledProcessError(p.returncode, command, output=p.stdout)
subprocess.CalledProcessError: Command '['ninja', '-w', 'dupbuild=err', '-d', 'explain']' returned non-zero exit status 1.

Stdout:
The Meson build system
Version: 0.49.999
Source dir: D:\a\1\s\test cases\unit\18 pkgconfig static
Build dir: C:\Users\VSSADM~1\AppData\Local\Temp\tmpi_umei5o
Build type: native build
Project name: pkg-config static
Project version: undefined
Native C compiler: clang (clang 7.0.1 "clang version 7.0.1 (tags/RELEASE_701/final)")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Message: /D/a/1/s/test cases/unit/18 pkgconfig static
Configuring foo.pc using configuration
Found pkg-config: D:\a\msys64\mingw64\bin\pkg-config.EXE (0.29.2)
Dependency foo found: YES 1.0
Build targets in project: 1
Found ninja-1.9.0 at 'D:\a\msys64\mingw64\bin\ninja.EXE'

ninja explain: deps for 'foomain@exe/main.c.obj' are missing
ninja explain: foomain@exe/main.c.obj is dirty
ninja explain: /a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a has no in-edge and is missing
ninja explain: /a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a is dirty
ninja explain: foomain.exe is dirty
ninja: error: '/a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a', needed by 'foomain.exe', missing and no known rule to make it


----------------------------------------------------------------------
Ran 216 tests in 164.934s

FAILED (errors=1, skipped=75)

Stdout:
The Meson build system
Version: 0.49.999
Source dir: D:\a\1\s\test cases\unit\18 pkgconfig static
Build dir: C:\Users\VSSADM~1\AppData\Local\Temp\tmpi_umei5o
Build type: native build
Project name: pkg-config static
Project version: undefined
Native C compiler: clang (clang 7.0.1 "clang version 7.0.1 (tags/RELEASE_701/final)")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Message: /D/a/1/s/test cases/unit/18 pkgconfig static
Configuring foo.pc using configuration
Found pkg-config: D:\a\msys64\mingw64\bin\pkg-config.EXE (0.29.2)
Dependency foo found: YES 1.0
Build targets in project: 1
Found ninja-1.9.0 at 'D:\a\msys64\mingw64\bin\ninja.EXE'

ninja explain: deps for 'foomain@exe/main.c.obj' are missing
ninja explain: foomain@exe/main.c.obj is dirty
ninja explain: /a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a has no in-edge and is missing
ninja explain: /a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a is dirty
ninja explain: foomain.exe is dirty
ninja: error: '/a/msys64/mingw64/x86_64-w64-mingw32/lib/libm.a', needed by 'foomain.exe', missing and no known rule to make it
```

NickeZ commented:
I can't reproduce the error locally. I installed msys2 and mingw-x86_64 and that test passes on my machine.

NickeZ commented:
I spoke to soon, it fails, but I get another error:

```
test_pkgconfig_static (__main__.AllPlatformTests) ... ERROR

Stdout:
The Meson build system
Version: 0.49.999
Source dir: C:/msys64/home/nicke/git/meson/test cases/unit/18 pkgconfig static
Build dir: C:/Users/nicke/AppData/Local/Temp/tmp5b1uajae
Build type: native build
Project name: pkg-config static
Project version: undefined
Native C compiler: clang (clang 7.0.1 "clang version 7.0.1 (tags/RELEASE_701/final)")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Message: /C/
Configuring foo.pc using configuration
Found pkg-config: C:\msys64\mingw64\bin/pkg-config.EXE (0.29.2)
WARNING: Static library 'foo' not found for dependency 'foo', may not be statically linked
Dependency foo found: YES 1.0
Build targets in project: 1
Found ninja-1.9.0 at 'C:\msys64\mingw64\bin/ninja.EXE'

[1/2] Compiling C object foomain@exe/main.c.obj.
FAILED: foomain@exe/main.c.obj
clang @foomain@exe/main.c.obj.rsp
../../../../../../msys64/home/nicke/git/meson/test cases/unit/18 pkgconfig static/main.c:1:10: fatal error: 'foo.h' file not found
#include <foo.h>
         ^~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
ninja explain: deps for 'foomain@exe/main.c.obj' are missing
ninja explain: foomain@exe/main.c.obj is dirty
ninja explain: foomain.exe is dirty


```

NickeZ commented:
yay, I fixed the bug on CI. The issue I had on my machine was unrelated.

TheQwertiest commented:
0.50?

NickeZ commented on mesonbuild/compilers/c.py:
```diff
@@ -935,18 +935,17 @@ def get_library_naming(self, env, libtype, strict=False):
-        if libtype in ('default', 'shared-static'):
-            patterns += self._get_patterns(env, prefixes, shlibext, True)
-            patterns += self._get_patterns(env, prefixes, stlibext, False)
+        if libtype == 'shared-static':
```
I don't have an opinion, I didn't add these names. I only removed `default` because I found it non-descriptive.

dcbaker commented:
I blamed it, the `static-shared`/`shared-static` madness is from lake 2017, so I guess I can't really ask to change it now, but those are terrible names, especially for dyslexics :(

NickeZ commented:
> I blamed it, the `static-shared`/`shared-static` madness is from late 2017, so I guess I can't really ask to change it now, but those are terrible names, especially for dyslexics :(

It is (still) an implementation detail, so I can change it if there is consensus on a better name. There is quite a high probability that something like this will sneak out into the public API and changing it now might nudge the potentially public API into a better direction.

NickeZ commented:
Added some docs

dcbaker commented:
I'll send a separate PR for the rename.

Test failures look unrelated BTW.
